### PR TITLE
[MERGE] tools: better placeholder support for _

### PR DESCRIPTION
**Pull request by mart-e**
Support two new syntaxes with `_()` method:

```py
_("This is so %s") % 2019
_("Text with %s placeholders", 42)
_("Text with %(num)s placeholders %(symbol)s", num=42, symbol="🧀")
```

The main benefit is to have a fallback on the source string if formating the
translated string fails (e.g. `"Foo %s"` translated `"Foo s%"`).

This PR converted most (but not all) calls to _ using placeholders to use the
new syntax. Hopefully, the remaining ones will be converted in the future and
avoid technical debt.

Use named placeholders when needed. If more than one placeholder is present in a
string, using named placeholder can be beneficial to reorder the placehodler
position (for languages like Japanese that do not use the same phrase
construction). On some sentences, it also brings clarity on the meaning of the
placeholders.

Task-id: 1853119

### CODE REVIEW

**Comment by rim-odoo** [comment]
:heart_eyes:

**Comment by mart-e** [comment]
Well still missing the second commit converting the 1240 calls to `_()` but I guess it's easy...

File: odoo/tools/translate.py [hunk 1]
```diff
@@ -443,8 +443,23 @@ def _get_lang(self, frame):
...
+    def __call__(self, source, *args, **kw):
+        translation = self._get_translation(source)
+        if args:
+            try:
+                return translation % args
+            except (TypeError, ValueError):
+                bad = translation
+                translation = source % args
+                _logger.error('Bad translation "%r" for "%r"', bad, source)
+        elif kw:
```
Comment (xmo-odoo) [inline odoo/tools/translate.py:455]: Could merge the two in getting the first non-emtpy of `args` and `kw`. Though we should probably also assert that we don't have *both* as that's nonsense.

**Comment by nim-odoo** [comment]
I think it's a great added value which avoids any overhead :+1: Moreover the API is similar to the one of `logging`.

File: odoo/tools/translate.py [hunk 2]
```diff
@@ -443,8 +443,23 @@ def _get_lang(self, frame):
...
         return lang
 
-    def __call__(self, source):
-        return self._get_translation(source)
+    def __call__(self, source, *args, **kw):
+        translation = self._get_translation(source)
+        if args:
+            try:
+                return translation % args
+            except (TypeError, ValueError):
```
Comment (xmo-odoo) [inline odoo/tools/translate.py:451]: Under what condition does this raise a ValueError?

File: odoo/tools/translate.py [hunk 3]
```diff
@@ -443,8 +443,23 @@ def _get_lang(self, frame):
...
+                bad = translation
+                translation = source % args
+                _logger.error('Bad translation "%r" for "%r"', bad, source)
+        elif kw:
+            try:
+                return translation % kw
+            except KeyError:
```
Comment (xmo-odoo) [inline odoo/tools/translate.py:458]: Also needs TypeError:
```python
>>> "%(a)d" % {'a': "xxx"}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: %d format: a number is required, not str
```

File: odoo/tools/translate.py (hunk 2 above)
Comment (mart-e) [reply odoo/tools/translate.py:451]:
```
>>> "Foo s%" % 42
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: incomplete format
```

Comment (xmo-odoo) [reply odoo/tools/translate.py:451]: Same error with a mapping, so really the only difference errors-wise is the mapping one can also raise a `KeyError`, which it's fine to catch in either case.

File: odoo/tools/translate.py [hunk 4]
```diff
@@ -443,8 +443,23 @@ def _get_lang(self, frame):
...
-        return self._get_translation(source)
+    def __call__(self, source, *args, **kw):
+        translation = self._get_translation(source)
+        if args:
+            try:
+                return translation % args
+            except (TypeError, ValueError):
+                bad = translation
+                translation = source % args
+                _logger.error('Bad translation "%r" for "%r"', bad, source)
```
Comment (xmo-odoo) [inline odoo/tools/translate.py:454]: Might be useful to use `logging.exception` or `exc_info=True` here. And some of the errors are related to incorrect values rather than an incorrect translation string.

Also `%r` does the quoting, so they're not necessary.

Comment (mart-e) [reply odoo/tools/translate.py]: Would it be useful to add the language in the error? (need a new call to `_get_lang`)

File: odoo/tools/translate.py [hunk 5]
```diff
@@ -443,8 +443,16 @@ def _get_lang(self, frame):
...
-        return self._get_translation(source)
+    def __call__(self, source, *args, **kw):
+        translation = self._get_translation(source)
+        if args or kw:
+            try:
+                return translation % (args or kw)
+            except (TypeError, ValueError, KeyError):
+                bad = translation
+                translation = source % (args or kw)
+                _logger.error('Bad translation %r for %r', bad, source)
```
Comment (xmo-odoo) [inline odoo/tools/translate.py:454]: How about
```suggestion
                _logger.error('Bad translation %r for %r', translation, source)
                return source % (args or kw)
```
Alternatively, if we want a single point of return, replace 450 by `translation = translation % (args or kw)`. And either way, lifting the error to right after the `except` clause seems like a good idea (the `bad` local seems wholly unnecessary).

File: odoo/tools/translate.py [hunk 6]
```diff
@@ -443,8 +443,16 @@ def _get_lang(self, frame):
                     lang = env['res.users'].context_get()['lang']
         return lang
 
-    def __call__(self, source):
-        return self._get_translation(source)
+    def __call__(self, source, *args, **kw):
+        translation = self._get_translation(source)
+        if args or kw:
```
Comment (xmo-odoo) [inline odoo/tools/translate.py:448]: Maybe add an error if `args` *and* `kw`? Especially given `kw` gets ignored entirely in that case.

File: odoo/tools/translate.py (hunk 5 above)
Comment (mart-e) [reply odoo/tools/translate.py:454]: I though about it but I wanted to log only if the issue was because of a bad translation or a bad source. If it fails in English as well, you shouldn't get a "Bar translation" error, no ?

Comment (xmo-odoo) [reply odoo/tools/translate.py:454]: My suggestion doesn't change that? It just moves things around a bit to avoid the unnecessary locale. If it fails in english as well then the error is not caught by either version, it just blows up.

File: odoo/tools/translate.py (hunk 4 above)
Comment (xmo-odoo) [reply odoo/tools/translate.py]: Maybe. Would that actually help? E.g. let's say you see such an error in the log, how would it get fixed, would you go to the translation list and look for the proper source / improper translation, or would you first filter by language then look for the proper source / improper translation?

Comment (mart-e) [reply odoo/tools/translate.py]: There is a good chance the admin knows what language are installed (except on the SaaS were you handle tons of dbs). And if not sure, can still make a global search on the db.

File: odoo/tools/translate.py (hunk 5 above)
Comment (mart-e) [reply odoo/tools/translate.py:454]:
```
2020-06-02 09:24:31,122 38822 ERROR mc odoo.tools.translate: Bad translation 'Text avec %s marqueur' for string 'Text with %s placeholder' 
Traceback (most recent call last):
  File "/home/mat/odoo/odoo/odoo/tools/translate.py", line 450, in __call__
    return translation % (args or kw)
TypeError: not all arguments converted during string formatting
```
The `Bad translation 'Text avec %s marqueur' for string 'Text with %s placeholder'` should not be there as the translation is actually correct, it's a programming error, not a translation issue.

File: odoo/tools/translate.py (hunk 4 above)
Comment (xmo-odoo) [reply odoo/tools/translate.py]: Either way it probably never hurts to have e.g. "Bad <lang> translation %r for %r" so might as well put it in if it's not too complicated.

Comment (xmo-odoo) [reply odoo/tools/translate.py]: Though maybe this could be done by modifying _get_translation to return both the translation and the lang, or delegate to a method which does that.

File: odoo/tools/translate.py (hunk 5 above)
Comment (xmo-odoo) [reply odoo/tools/translate.py]: Well yes, all I'm saying is I don't think my suggestion changes that, it's already there in the original no?

File: odoo/tools/translate.py [hunk 7]
```diff
@@ -443,8 +443,17 @@ def _get_lang(self, frame):
...
+    def __call__(self, source, *args, **kw):
+        translation = self._get_translation(source)
+        assert not (args and kw)
+        if args or kw:
+            try:
+                return translation % (args or kw)
+            except (TypeError, ValueError, KeyError):
+                bad = translation
+                translation = source % (args or kw)
+                _logger.exception('Bad translation %r for string %r', bad, source, exc_info=True)
```
[the remark this answers is not in the document]
Comment (rco-odoo) [reply odoo/tools/translate.py]: Now I understand your comment to Xavier's proposal.
If `source % (args or kw)` fails, you don't log and just raise the expected error on the source string.

File: odoo/tools/translate.py [hunk 8]
```diff
@@ -443,8 +443,18 @@ def _get_lang(self, frame):
...
+    def __call__(self, source, *args, **kwargs):
+        translation = self._get_translation(source)
+        assert not (args and kwargs)
+        if args or kwargs:
+            try:
+                return translation % (args or kwargs)
+            except (TypeError, ValueError, KeyError):
+                bad = translation
+                # fallback on source
+                translation = source % (args or kwargs)
```
Comment (rco-odoo) [inline odoo/tools/translate.py:455]:
```suggestion
                # fallback: apply to source before logging exception (in case source fails)
                translation = source % (args or kwargs)
```

File: odoo/addons/test_translation_import/tests/test_term_count.py [hunk 9]
```diff
@@ -269,6 +269,72 @@ def test_export_pollution(self):
...
+        translation = self.env['ir.translation'].create({
+            'src': 'Text with %s placeholder',
+            'value': 'Text avec %s marqueur',
+            'type': 'code',
+            'name': 'addons/test_translation_import/tests/test_count_term.py',
+            'res_id': 0,
+            'lang': 'fr_BE',
+        })
+
+        context = {'lang': "fr_BE"}
```
Comment (rco-odoo) [inline odoo/addons/test_translation_import/tests/test_term_count.py:286]: Move that line at the beginning of method.

File: odoo/addons/test_translation_import/tests/test_term_count.py [hunk 10]
```diff
@@ -269,6 +269,72 @@ def test_export_pollution(self):
...
+        # correctly translate
+        self.assertEqual(
+            _("Text with %s placeholder", 1),
+            "Text avec 1 marqueur",
+            "Translation placeholders were not applied"
+        )
+
+        # too many arguments should raise an error
```
Comment (rco-odoo) [inline odoo/addons/test_translation_import/tests/test_term_count.py:294]:
```suggestion
        # source error: wrong arguments
```
Too bad you cannot `assertNoLogs()` here :wink:

File: odoo/addons/test_translation_import/tests/test_term_count.py [hunk 11]
```diff
@@ -269,6 +269,72 @@ def test_export_pollution(self):
...
+        with self.assertRaises(TypeError), self.cr.savepoint():
+            _("Text with %s placeholder", 1, "🧀")
+
+        # wrong translation should fallback on source
```
Comment (rco-odoo) [inline odoo/addons/test_translation_import/tests/test_term_count.py:298]:
```suggestion
        # translation error: log error and fallback on source
```

File: odoo/addons/test_translation_import/tests/test_term_count.py [hunk 12]
```diff
@@ -269,6 +269,72 @@ def test_export_pollution(self):
...
+        })
+
+        # correctly translate
+        self.assertEqual(
+            _("Text with %(num)s placeholders %(symbol)s", num=2, symbol="🧀"),
+            "Text avec 2 marqueurs 🧀",
+            "Translation placeholders were not applied"
+        )
+
+        # not enough arguments should raise an error
```
Comment (rco-odoo) [inline odoo/addons/test_translation_import/tests/test_term_count.py:325]:
```suggestion
        # source error: wrong arguments
```

File: odoo/addons/test_translation_import/tests/test_term_count.py [hunk 13]
```diff
@@ -269,6 +269,72 @@ def test_export_pollution(self):
...
+        with self.assertRaises(KeyError), self.cr.savepoint():
+            _("Text with %(num)s placeholders %(symbol)s", symbol="🧀")
+
+        # wrong translation should fallback on source
```
Comment (rco-odoo) [inline odoo/addons/test_translation_import/tests/test_term_count.py:329]:
```suggestion
        # translation error: log error and fallback on source
```

**Comment by Yenthe666** [comment]
@mart-e I guess this can now use a doc update too?

**Comment by mart-e** [comment]
> I guess this can now use a doc update too?
@Yenthe666 there is a doc update in this PR. Or are you talking about another one?

**Comment by Yenthe666** [comment]
Cool, I missed that one. Awesome work 👏

File: doc/reference/guidelines.rst [hunk 14]
```diff
@@ -869,32 +874,31 @@ the form ``_('literal string')`` and nothing else:
     # and the following will of course not work as already explained:
     error = _("Product %s is out of stock!" % product.name)
 
-    # bad: field values are automatically translated by the framework
-    # This is useless and will not work the way you think:
-    error = _("Product %s is not available!") % _(product.name)
-    # and the following will of course not work as already explained:
-    error = _("Product %s is not available!" % product.name)
-
```
Comment (rco-odoo) [inline doc/reference/guidelines.rst:877]: Why did you remove this?  Those examples are still bad!

File: doc/reference/translations.rst [hunk 15]
```diff
@@ -135,16 +136,16 @@ Plural
 ^^^^^^
 **Don't** pluralize terms the English-way::
 
-    msg = _("You have %s invoice") % invoice_count
+    msg = _("You have %(count)s invoice, count=invoice_count)
```
Comment (rco-odoo) [inline doc/reference/translations.rst:139]:
```suggestion
    msg = _("You have %(count)s invoice", count=invoice_count)
```

File: doc/reference/guidelines.rst (hunk 14 above)
Comment (mart-e) [reply doc/reference/guidelines.rst:877]: it's actually duplicated, the exact same examples are just above

**Comment by mart-e** [comment]
> What about the export of terms? Will `_("...", ...)` be handled the same way as `_("...")`?

Yes, babel handles them fine. I tried exporting test_translation_import module and it contained all the examples.

**Comment by mart-e** [comment]
@robodoo r+ merge

**Comment by nim-odoo** [comment]
@mart-e it might need some internal communication so people know about it.

**Comment by AntoineVDV** [comment]
> @mart-e it might need some internal communication so people start using it.

And doc 👀

cc @Feyensv if you want to open a 468th branch

**Comment by mart-e** [comment]
> And doc eyes

@AntoineVDV you are the second one that assumes we haven't written doc and don't look at the diff to check :thinking:

**Comment by Yenthe666** [comment]
@mart-e actually I looked at the diff (and I assume @AntoineVDV also did). It was a pretty big one so I/we missed it.
Besides, let's not forget the times there where nearly zero docs, we got too used to it 😉

**Comment by AntoineVDV** [comment]
> > And doc eyes
> 
> @AntoineVDV you are the second one that assumes we haven't written doc and don't look at the diff to check

Good then ! 👍
I actually looked for .rst files but I guess that I did before the entire diff was loaded 😅
